README
This plasmoid was inspired by the KBirthday kicker
applet for KDE3, I've been using for years now.
Making a slow transition to KDE4 now, that was something
I was missing.
It reminds you at the birthdays and anniversaries of your
loved ones (in advance).

To install:

make sure, you have the following packages installed:
    -kde4pimlibs-devel
    -libqt4-devel
    -libkde4-devel
    -cmake
*Note: The names of the packages are take from openSUSE.
       They might differ from the package names in
       your distribution.

Unpack the archive and cd into the directory created.
Then issue the following command on the command line:

=====================================================
cd build
cmake -DCMAKE_INSTALL_PREFIX=`kde4-config --prefix` ../
make 
sudo make install
=====================================================

If no 'kde4-config' exists on your installation, try
'kde-config' instead. If that fails as well, try to 
locate the installation path of your KDE4. If e.g. 
KDE4 is installed in /usr/kde4, replace the above command
with 
    cmake -DCMAKE_INSTALL_PREFIX=/usr ..

You might have to issue
  kdequitapp plasma
  plasma
in order to add the plasmoid to the list of available plasmoids

The addressbook.vcf file provided is containing test users.
It's quite useful to add them to KAddressbook of a seperate
account if one wants to test the applet. It contains quite a
few test cases around year change



ChangeLog
20100111
        -============= V 0.9.73 ==================
        -French translation added (Alain Portal)
        -fixing a couple of misspellings
        -applied patches provided by Alain Portal
20090306
        -============= V 0.9.72 ==================
        -I18N files added
        -added German language file
20090303
        -============= V 0.9.71 ==================
        -applied openSUSE patches
        -adapted cmake build files to KDE 4.2 final
        -crash when started in hour after midnight - FIXED
20081223
        -============= V 0.9.70 ==================
        -I18N_ARGUMENT_MISSING  FIXED by Patch from Tumsa on KDE-Looks org
        -Since the ToolTip interface changed from KDE 4.1 to 4.2 this version depends on 4.2
        -Extended panel support - this is not final yet!
20081026
        -============= V 0.8.0 ==================
        -if added to a panel instead of the desktop, number of events will be shown
        -added tooltip if added to panel (tooltip has yet to be extended)
         tooltip interface keeps chaning - KDE >4.1.70 supported only at the moment
        -coming events in next year and past in events in last year were not displayed (FIXED)
        -count of coming events are calculated in order to resize plasmoid 
         /resizing itself not working yet/
        -Added Icon	
20081023
        -============= V 0.7.1 ==================
        -updates on midnight
        -sorting list of events - closest events first
        -using the font colour for headings (just a bit darker)
        -Settings now stored when the config dialog is closed
20081021  
        -============= V 0.7 ==================
        -initial Release



src/applet/blistentry.h
#ifndef BLISTENTRY_H
#define BLISTENTRY_H

/* ************************************************
 *  @name: blistentry.h 
 *  @author: Meinhard Ritscher
 *  @date: 2008-10-21
 *
 *  $Id:  $
 *
 *  Description
 *  ============
 *  Data object storing all necessary info around
 *  the event
 *
 *  Histrory
 *  ============
 *
 * ********************************************** */
/* *************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 or (at your option)    *
 *   any later version.                                                    *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY of FITNESS FOR A PARTICULAR PURPOSE. See the GNU      *
 *   General Public Licencse for more details.                             *
 *                                                                         *
 *   A copy of the license should be part of the package. If not, you can  *
 *   find it here: http://www.gnu.org/licenses/gpl.html                    *
 *                                                                         *
 ***************************************************************************/

#include <QString>
#include <QDate>

#include <KDebug>


class BListEntry{

    public:
        BListEntry(const QString &name, const QDate &date);
        ~BListEntry(){}

        QString name() const { return m_name; }
        unsigned int age() const{ return m_age;}
        int remainingDays() const{ return m_remainingDays;}
        void calculateDays();
        static bool lessThan(const BListEntry *a, const BListEntry *b);

    private:
    /** @brief The formatted name of that person. */
        QString m_name;
    /** @brief How old turns the person / the anniversary. */
        unsigned int m_age;
    /** @brief How many days until the event (this might be negative). */
        int m_remainingDays;

    /** @brief The date of the event*/
        QDate m_date;
};

#endif


src/applet/blistentry.cpp
#include "blistentry.h"

/* ************************************************
 *  @name: blistentry.cpp 
 *  @author: Meinhard Ritscher
 *
 *  $Id:  $
 *
 *  See header file for description and history
 *
 * ********************************************** */
/* *************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 or (at your option)    *
 *   any later version.                                                    *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY of FITNESS FOR A PARTICULAR PURPOSE. See the GNU      *
 *   General Public Licencse for more details.                             *
 *                                                                         *
 *   A copy of the license should be part of the package. If not you can   *
 *   find it here: http://www.gnu.org/licenses/gpl.html                    *
 *                                                                         *
 ***************************************************************************/


BListEntry::BListEntry(const QString &name, const QDate &date)
    :m_name(name)
{
    m_date = date;
    calculateDays();
}

void BListEntry::calculateDays()
{
    QDate today = QDate::currentDate();
//    today = QDate(2001,2,2);
    int year = today.year();
    int month = today.month();

// every event from the past six months
// are calculated in the past, events of the next
// six months will be in the future
    int deltaMonth = month - m_date.month();

    // if the year has turned in the past six or
    // will turn in the next six months, we have
    // to adust the year
    if(deltaMonth < -5){
        year--;
    }else if(deltaMonth > 5){
        year++;
    }
    m_age = year - m_date.year();
    QDate tempDate = QDate(year, m_date.month(), m_date.day());
    m_remainingDays = today.daysTo(tempDate);
//    kDebug() << m_name << " - " << m_remainingDays;
}

bool BListEntry::lessThan(const BListEntry *a, const BListEntry *b)
{
    return ( a->remainingDays() == b->remainingDays())?
                (a->age() < b->age()):(a->remainingDays() < b->remainingDays());
}


src/applet/kbirthdayview.h
#ifndef KBIRTHDAY_VIEW_H
#define KBIRTHDAY_VIEW_H

/* ************************************************
 *  @name: kbirthdayview.h 
 *  @author: Meinhard Ritscher
 *  @date: 2008-12-05
 *
 *  $Id:  $
 *
 *  Description
 *  ============
 *  Displays the Birthday and Anniversary events
 *  when plasmoid is added to a panel
 *  This is more or less the code used from the
 *  device notifier plasmoid
 *
 *  Histrory
 *  ============
 *  2008-12-05 created
 *
 * ********************************************** */
/*  Copyright 2007 by Alexis Ménard <darktears31@gmail.com>

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Library General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Library General Public License for more details.

    You should have received a copy of the GNU Library General Public License
    along with this library; see the file COPYING.LIB.  If not, write to
    the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
    Boston, MA 02110-1301, USA.
*/

// Qt
#include <QTreeView>

class QModelIndex;

//namespace KBirthday
//{
  /**
  * @short The view used to display the single events
  *
  */
  class KBirthdayView : public QTreeView
  {

  Q_OBJECT
    public:
        /**
        * Constructor of view
        * @param parent the parent of this object
        **/
        KBirthdayView(QWidget *parent = 0);

        /**
        * Default destructor
        **/
        virtual ~KBirthdayView();

        /**
        * Creates rects in widget coordinates for the model
        */
        void calculateRects();

    protected:
        /**
        * Call when the view is resized
        * @param event the resize event
        **/
        void resizeEvent(QResizeEvent * event);

        /**
        * Call when a mouse move event is catch
        * @param event the mouse event
        **/
        void mouseMoveEvent(QMouseEvent *event);

        /**
        * Call when a mouse move event is catch
        * @param event the mouse event
        **/
        void mousePressEvent(QMouseEvent *event);

        /**
        * Call when cursor leave the widget
        * @param event the leave event
        **/
        void leaveEvent(QEvent *event);

        /**
        * Move the cursor in the way describe by cursorAction
        * @param cursorAction the cursor action
        **/
        QModelIndex moveCursor(CursorAction cursorAction,Qt::KeyboardModifiers );

        /**
        * Call when the view is paint
        * @param event the paint event
        **/
        void paintEvent(QPaintEvent *event);

        /**
        * Paint a header item
        * @param painter the painter used to paint
        * @param itemRect the rect where the item will be paint
        * @param index the QModelIndex that represent the item to paint
        **/
        void paintHeaderItem(QPainter &painter,const QRect &itemRect,const QModelIndex &index);

        /**
        * Paint an item in the view by using the delegate
        * @param painter the painter used to paint
        * @param itemRect the rect where the item will be paint
        * @param index the QModelIndex that represent the item to paint
        **/
        void paintItem(QPainter &painter,const QRect &itemRect,const QModelIndex &index);

        /**
        * Return an index at the position "point" if exist otherwise return a default
        * constructed value of QModelIndex
        * @param point the point where we will looking for an item
        **/
        QModelIndex indexAt(const QPoint& point) const;

        static const int HEADER_LEFT_MARGIN = 5;
        static const int HEADER_TOP_MARGIN = 5;
        static const int HEADER_HEIGHT = 35;
        static const int COLUMN_EJECT_MARGIN = 5;
        static const int COLUMN_EJECT_SIZE = 50;
        static const int TOP_OFFSET = 5;

    private:
        ///The hovered index
        QPersistentModelIndex m_hoveredIndex;
        QHash<QModelIndex,QRect> m_itemRects;
  };

//} //namespace KBirthday
#endif


src/applet/kbirthdayapplet.h
#ifndef KBIRTHDAY_APPLET_H
#define KBIRTHDAY_APPLET_H

/* ************************************************
 *  @name: addressbook.h 
 *  @author: Meinhard Ritscher
 *  @date: 2008-10-21
 *
 *  $Id:  $
 *
 *  Description
 *  ============
 *
 *
 *  Histrory
 *  ============
 *
 * ********************************************** */
/* *************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 or (at your option)    *
 *   any later version.                                                    *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY of FITNESS FOR A PARTICULAR PURPOSE. See the GNU      *
 *   General Public Licencse for more details.                             *
 *                                                                         *
 *   A copy of the license should be part of the package. If not you can   *
 *   find it here: http://www.gnu.org/licenses/gpl.html                    *
 *                                                                         *
 ***************************************************************************/

#include "ui_kbirthdayapplet_config.h"

//own
#include "kbirthdaydialog.h"

//Plasma
#include <Plasma/PopupApplet>
#include <Plasma/Svg>
#include <Plasma/DataEngine>

class QSizeF;
class BListEntry;

namespace Plasma{

    class Svg;

}

//namespace KBirthday
//{
    class KBirthdayApplet : public Plasma::PopupApplet
    {
    Q_OBJECT
    
    friend class KBirthdayDialog;
    
        public:
            KBirthdayApplet(QObject *parent, const QVariantList &args);
            ~KBirthdayApplet();

            void paintInterface(QPainter *painter,
                                const QStyleOptionGraphicsItem *option,
                                const QRect& contentsRect);
            /**
            * initialize the applet (called by plasma automatically)
            */
            void init();

            /**
            * The widget that displays the list of devices.
            */
            QWidget *widget();

        protected:
            void createConfigurationInterface(KConfigDialog *parent);
            void constraintsEvent(Plasma::Constraints constraints);
            void popupEvent(bool show);

        public slots:
            void dataUpdated(const QString &name, const Plasma::DataEngine::Data &data);
            /**
            * @internal Sets the tooltip content properly before showing.
            */
            void toolTipAboutToShow();
            /**
            * @internal Clears memory when needed.
            */
            void toolTipHidden();

        protected slots:
            void configAccepted();
            
        private:
            void updateEventList(const QList <QVariant> &anniversaries, QList<BListEntry*>** eventList);
            void updateEventCount(void);
            bool testThreshold(const int remainingDays);
            int printEvent(QPainter* p, int x, int y, int width, const BListEntry* entry);
            QString remainingDaysString(const int remainingDays);

        private:
        /** @brief The configuration dialog. */
            Ui::KBirthdayAppletConfig m_ui;

        /** @brief Various configurable settings. */
            QColor m_colour;
            QColor m_past_colour;
            QColor m_highlight_colour;

            int m_highlightTreshold;
            int m_eventTreshold;
            int m_pastTreshold;
            bool m_showAnniversaries;
        /** @brief  The dialog where events are displayed when in panel.*/
            KBirthdayDialog* m_pDialog;
        /** @brief  Set to true if plasmoid lives on a panel.*/
            bool m_isOnPanel;
        /** @brief A pointer to address book dataengine. */
            Plasma::DataEngine *m_pKabcEngine;
            Plasma::DataEngine *m_pAkonadiEngine;
        /** @brief A pointer to the list of birthdays. */
            QList<BListEntry*>* m_pBirthdays;
        /** @brief A pointer to the lsit of anniversaries. */
            QList<BListEntry*>* m_pAnniversaries;
        /** @brief number of outstanding events */
            int m_eventCount;
        /** @brief number of outstanding events */
            int m_birthEventCount;
        /** @brief number of outstanding events */
            int m_anniEventCount;
        /** @brief the icon used when the applet is in the taskbar */
            Plasma::IconWidget *m_pIcon;
        /** @brief Background when displayed in a panel */
            Plasma::Svg* m_svg;
    };
//} // namespace KBrithday

#endif


src/applet/kbirthdaydialog.h
#ifndef KBIRTHDAY_DIALOG_H
#define KBIRTHDAY_DIALOG_H

/* ************************************************
 *  @name: kbirthdaydialog.h 
 *  @author: Meinhard Ritscher
 *  @date: 2008-12-05
 *
 *  $Id:  $
 *
 *  Description
 *  ============
 *
 *
 *  Histrory
 *  ============
 *
 * ********************************************** */
/* *************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 or (at your option)    *
 *   any later version.                                                    *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY of FITNESS FOR A PARTICULAR PURPOSE. See the GNU      *
 *   General Public Licencse for more details.                             *
 *                                                                         *
 *   A copy of the license should be part of the package. If not you can   *
 *   find it here: http://www.gnu.org/licenses/gpl.html                    *
 *                                                                         *
 ***************************************************************************/

//Qt
#include <QWidget>

//Qt
class QLabel;
class QStandardItemModel;

//own
class KBirthdayApplet;


//namespace KBirthday
//{
  class KBirthdayView;

  /**
  * @short The dialog used to display events in a popup
  *
  */
  class KBirthdayDialog : public QObject
  {
  Q_OBJECT
      public:
          /**
          * Constructor of the dialog
          * @param notifier the birthday applet attached to this dialog
          * @param parent the parent of this object
          **/
          KBirthdayDialog(KBirthdayApplet* applet,QObject* parent = 0);
          
          /**
          * Default destructor
          **/
          virtual ~KBirthdayDialog();

         /**
          * Returns the related QWidget.
          **/
          QWidget* dialog();

          /**
          * Show the dialog
          **/
          void show();

          /**
           * Updates the event list
           **/
           void updateEventList();

      private:
         /**
          * @internal build the dialog
          **/
          void buildDialog();

      private:
          /** @brief The applet attached to this item */
          KBirthdayApplet* m_pApplet;
          /** @brief The widget which displays the events*/
          QWidget* m_pWidget;
          /** @brief QLabel which represent the title */
          QLabel* m_pLabel;
          /** @brief  */
          KBirthdayView *m_pEventView;
          /** @brief  */
          QStandardItemModel* m_pEventModel;
  };
  
#endif


src/applet/kbirthdayapplet.cpp
#include "kbirthdayapplet.h"

/* ************************************************
 *  @name: kbirthdayapplet.cpp 
 *  @author: Meinhard Ritscher
 *
 *  $Id:  $
 *
 *  See header file for description and history
 *
 * ********************************************** */
/* *************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 or (at your option)    *
 *   any later version.                                                    *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY of FITNESS FOR A PARTICULAR PURPOSE. See the GNU      *
 *   General Public Licencse for more details.                             *
 *                                                                         *
 *   A copy of the license should be part of the package. If not you can   *
 *   find it here: http://www.gnu.org/licenses/gpl.html                    *
 *                                                                         *
 ***************************************************************************/

// Qt
#include <QPainter>
#include <QFontMetrics>
#include <QSizeF>
#include <QPair>
#include <QTime>

// KDE
#include <KConfigDialog>
#include <KDebug>

// Plasma
#include <Plasma/Theme>
#include <Plasma/Svg>
#include <Plasma/IconWidget>
#if KDE_IS_VERSION(4,1,80)
#include <Plasma/ToolTipManager>
#endif

//own
#include "blistentry.h"

// linking this class to the desktop file so plasma can load it
K_EXPORT_PLASMA_APPLET(kbirthdayapplet, KBirthdayApplet)


typedef QPair<QString, QDate> KabEntry;
Q_DECLARE_METATYPE(KabEntry)


static const char *BIRTHDAY_SOURCE    = "Birthdays";
static const char *ANNIVERSARY_SOURCE = "Anniversaries";
static const char *UTC_SOURCE         = "UTC";
static const char *COLLECTION_SOURCE  = "ContactCollections";

KBirthdayApplet::KBirthdayApplet(QObject *parent, const QVariantList &args)
    : Plasma::PopupApplet(parent, args),
    m_pDialog(0),
    m_isOnPanel(false),
    m_pKabcEngine(0),
    m_pAkonadiEngine(0),
    m_pBirthdays(0),
    m_pAnniversaries(0),
    m_eventCount(0),
    m_birthEventCount(0),
    m_anniEventCount(0),
    m_svg(0)
{
    kDebug() << "INIT init INIT init INIT";
    setBackgroundHints(DefaultBackground);
    setAspectRatioMode(Plasma::IgnoreAspectRatio);
    setHasConfigurationInterface(true);
    resize(250, 300);
    
    // let's initialize the widget
    (void)widget();
    
    //m_pIcon = new Plasma::IconWidget(KIcon("cookie",NULL), QString());
    //setPopupIcon(m_pIcon->icon());
}

KBirthdayApplet::~KBirthdayApplet()
{
    if (hasFailedToLaunch()) {
        //Do some cleanup here
    } else {
    }
    delete m_pDialog;
    delete m_pBirthdays;
    delete m_pAnniversaries;
}

void KBirthdayApplet::popupEvent(bool show)
{
    if (show && m_isOnPanel) {
#if KDE_IS_VERSION(4,1,80)
        Plasma::ToolTipManager::self()->clearContent(this);
#endif
    }
}

void KBirthdayApplet::init()
{
    KConfigGroup configGroup = config();
    m_colour      = configGroup.readEntry("Color",                QColor(Qt::white));
    m_past_colour = configGroup.readEntry("Past Color",           QColor(150,150,150));
    m_highlight_colour = configGroup.readEntry("Highlight Color", QColor(170,0,0));
    m_highlightTreshold = configGroup.readEntry("Highlight Treshold", 2);
    m_eventTreshold = configGroup.readEntry("Event Treshold",       30);
    m_pastTreshold = configGroup.readEntry("Past Treshold",           2);
    m_showAnniversaries = configGroup.readEntry("Show Anniversaries", true);

    m_pKabcEngine = dataEngine("kabc");
    if (m_pKabcEngine) {
        m_pKabcEngine->connectSource(BIRTHDAY_SOURCE, this);
        m_pKabcEngine->connectSource(ANNIVERSARY_SOURCE, this);
    } else {
        //kDebug() << "Could not load kabc dataEngine";
        setFailedToLaunch(true, "Could not load kabc dataEngine");
    }
    Plasma::DataEngine* timeEngine = dataEngine("time");
    if (timeEngine) {
        timeEngine->connectSource("UTC", this, 360000, Plasma::AlignToHour);
    } else {
        kDebug() << "Warning: Could not load time dataEngine - no nightly update possible";
    }
    m_pAkonadiEngine = dataEngine("akonadi");
    if (m_pAkonadiEngine) {
        m_pAkonadiEngine->connectSource(COLLECTION_SOURCE, this);
    } else {
        kDebug() << "Could not load Akonadi dataEngine";
       // setFailedToLaunch(true, "Could not load kabc dataEngine");
    }

    m_svg = new Plasma::Svg(this);
    m_svg->setImagePath("widgets/birthdaycake");
    m_svg->setContainsMultipleImages(false);
 
}

QWidget *KBirthdayApplet::widget()
{
    if (!m_pDialog) {
        m_pDialog = new KBirthdayDialog(this);
        kDebug() << "KBirthdayDialog created";
    }
    return m_pDialog->dialog();
}

void KBirthdayApplet::constraintsEvent(Plasma::Constraints constraints)
{
    if (constraints & Plasma::FormFactorConstraint){
        Plasma::FormFactor formFctr = formFactor();
        if (formFctr == Plasma::Horizontal || formFctr == Plasma::Vertical) {
            setMinimumSize(QSize(42,42));
            setAspectRatioMode(Plasma::ConstrainedSquare);
#if KDE_IS_VERSION(4,1,80)
            Plasma::ToolTipManager::self()->registerWidget(this);
#endif
            setAspectRatioMode(Plasma::IgnoreAspectRatio);
            m_isOnPanel = true;
            kDebug() << "PANEL PANEL PANEL";

        } else if (formFctr == Plasma::MediaCenter || formFctr == Plasma::Planar) {
#if KDE_IS_VERSION(4,1,80)
            Plasma::ToolTipManager::self()->unregisterWidget(this);
#endif
            m_isOnPanel = false;
        }
    }
}

void KBirthdayApplet::paintInterface(QPainter *p,
    const QStyleOptionGraphicsItem *option, const QRect &contentsRect)
{
    Q_UNUSED(option)

    const int left   = contentsRect.left();
    const int top    = contentsRect.top();
    const int width  = contentsRect.width();

    p->save();

    // if the widget lives in a panel, just display the number of comming events
    if (m_isOnPanel) {
    
        m_svg->resize(contentsRect.size());
        if (m_svg->hasElement("Brownie")) {
            m_svg->paint(p, contentsRect, "Brownie");
        } else {
            kDebug() << "Brownie not found in Svg. ...";
        }

        QFont font(p->font());
        font.setPointSize(font.pointSize() + 2);
        font.setBold(true);
        p->setFont(font);
        p->setPen(m_colour);

        p->drawText(contentsRect,Qt::AlignCenter|Qt::AlignVCenter,QString("%1").arg(m_eventCount));

        p->restore();
        return;
    }

    int entryHight = QFontMetrics(p->font()).height()*2;
    setMinimumSize(QSize((int)contentsRect.left()*2+250,
                    (int)contentsRect.top()*2+entryHight*m_eventCount-3 ));
    int top_offset = top+2;

    p->setPen(m_colour.darker());

    p->fillRect(left, top_offset, width, 1, m_colour.darker());
    top_offset += 2;

    p->setRenderHint(QPainter::SmoothPixmapTransform);
    p->setRenderHint(QPainter::Antialiasing);

    if (   (!m_pBirthdays     || !m_birthEventCount) 
        && (!m_pAnniversaries || !m_anniEventCount))
    {
        p->drawText(left, top_offset, width, p->fontMetrics().height(),
                         Qt::AlignCenter, i18n("No Events"));
    } else {

        if (m_pBirthdays && m_birthEventCount) {
            p->drawText(left, top_offset, width, p->fontMetrics().height(),
                            Qt::AlignCenter, i18n("Birthdays"));
            top_offset += p->fontMetrics().height();
            foreach(const BListEntry *entry, *m_pBirthdays){
                top_offset = printEvent(p, left, top_offset, width, entry) + 3;
            }
        }
        if (m_showAnniversaries && m_anniEventCount) {
            p->fillRect(left, top_offset, width, 1, m_colour.darker());
            top_offset += 2;

            p->drawText(left, top_offset, width, p->fontMetrics().height(),
                            Qt::AlignCenter, i18n("Anniversaries"));
            top_offset += p->fontMetrics().height();
            foreach(const BListEntry *entry, *m_pAnniversaries){
                top_offset = printEvent(p, left, top_offset, width, entry) + 3;
            }
        }
    }
    p->restore();
}

void KBirthdayApplet::toolTipAboutToShow()
{
//    kDebug() << KDE::versionString() << "toolTipAboutToShow";
#if KDE_IS_VERSION(4,1,80)
// Tooltip handling seems to have changed a lot between 
// KDE version 4.1 and 4.2
    Plasma::ToolTipContent toolTip;
    toolTip.setMainText(i18n("Birthday Reminder"));
    if (m_eventCount > 0)
        toolTip.setSubText(i18np("One outstanding event", "%1 outstanding events", m_eventCount ));
    else
        toolTip.setSubText(i18n("No outstanding event"));
    Plasma::ToolTipManager::self()->setContent(this, toolTip);
#endif
}

void KBirthdayApplet::toolTipHidden()
{
#if KDE_IS_VERSION(4,1,80)
    Plasma::ToolTipManager::self()->clearContent(this);
#endif
}


void KBirthdayApplet::createConfigurationInterface(KConfigDialog *parent)
{
    QWidget *widget = new QWidget;

    m_ui.setupUi(widget);

    parent->setButtons(KDialog::Ok | KDialog::Cancel);
    parent->addPage(widget, parent->windowTitle(), icon());

    m_ui.colour->setColor(m_colour);
    m_ui.pastColour->setColor(m_past_colour);
    m_ui.highlightColour->setColor(m_highlight_colour);

    m_ui.spinEventTreshold->setValue(m_eventTreshold);
    m_ui.spinPast->setValue(m_pastTreshold);
    m_ui.spinHighlight->setValue(m_highlightTreshold);

    m_ui.showAnniversaries->setChecked(m_showAnniversaries);

    connect(parent, SIGNAL(okClicked()), this, SLOT(configAccepted()));
}

/*
 * This method is called when the user clicks ok
 * in the config dialog
 *
 */

void KBirthdayApplet::configAccepted()
{

    m_colour            = m_ui.colour->color();
    m_past_colour       = m_ui.pastColour->color();
    m_highlight_colour  = m_ui.highlightColour->color();

    m_eventTreshold     = m_ui.spinEventTreshold->value();
    m_pastTreshold      = m_ui.spinPast->value();
    m_highlightTreshold = m_ui.spinHighlight->value();

    m_showAnniversaries = m_ui.showAnniversaries->isChecked();

    KConfigGroup configGroup = config();
    configGroup.writeEntry("Color",          m_colour);
    configGroup.writeEntry("Past Color", m_past_colour);
    configGroup.writeEntry("Highlight Color", m_highlight_colour);
    configGroup.writeEntry("Highlight Treshold",m_highlightTreshold);
    configGroup.writeEntry("Event Treshold",m_eventTreshold);
    configGroup.writeEntry("Past Treshold",m_pastTreshold);
    configGroup.writeEntry("Show Anniversaries",m_showAnniversaries);

    updateEventCount();
    update();

    emit configNeedsSaving();
}

void KBirthdayApplet::dataUpdated(const QString &name, const Plasma::DataEngine::Data &data)
{
    if (QString::compare(name, COLLECTION_SOURCE) == 0){
        kDebug() << "Collection updated";
    }else if (QString::compare(name, BIRTHDAY_SOURCE) == 0){
        updateEventList(data["Birthdays"].toList(), &m_pBirthdays);
        updateEventCount();
    }else if (QString::compare(name, ANNIVERSARY_SOURCE) == 0){
        updateEventList(data["Anniversaries"].toList(), &m_pAnniversaries);
        updateEventCount();
    }else if (QString::compare(name, UTC_SOURCE) == 0){
        // if midnight
        if ( QTime::currentTime().hour() == 0 ) {
	    if(m_pBirthdays){
	    //if started round midnight thees might not have
	    //created yet
            foreach(BListEntry *entry, *m_pBirthdays) {
                entry->calculateDays();
            }
	    }
	    if(m_pAnniversaries){
            foreach(BListEntry *entry, *m_pAnniversaries) {
                entry->calculateDays();
            }
        }
      }
    }
    update();
}

void KBirthdayApplet::updateEventList(const QList<QVariant> &list, QList<BListEntry*>** eventList)
{
    delete *eventList;
    *eventList = new QList<BListEntry*>();
    foreach(const QVariant &person, list){
        KabEntry ke = person.value<KabEntry>();
        BListEntry *entry = new BListEntry(ke.first, ke.second);
        (*eventList)->append(entry);
    }
    qSort((*eventList)->begin(), (*eventList)->end(), BListEntry::lessThan);
}

void KBirthdayApplet::updateEventCount()
{
    m_eventCount      = 0;
    m_birthEventCount = 0;
    m_anniEventCount  = 0;
    if (m_pBirthdays && ! m_pBirthdays->isEmpty()) {
        foreach(const BListEntry *entry, *m_pBirthdays) {
            if ( testThreshold(entry->remainingDays())) {
                    m_birthEventCount++;
                    // kDebug() << "Found birthday event" << entry->name();
            }
        }
    }
    if (m_pAnniversaries && m_showAnniversaries && ! m_pAnniversaries->isEmpty()) {
        foreach(const BListEntry *entry, *m_pAnniversaries){
            if ( testThreshold(entry->remainingDays())) {
                   m_anniEventCount++;
                   // kDebug() << "Found anniv event" << entry->name();
            }
        }
    }
    m_eventCount = m_birthEventCount + m_anniEventCount;
    // kDebug() << "EventCount now " << m_eventCount << " : " << m_birthEventCount << " : " <<  m_anniEventCount;
    
    if (m_pDialog)
      m_pDialog->updateEventList();
}

bool KBirthdayApplet::testThreshold(const int remainingDays)
{
  return ( (remainingDays > 0 && remainingDays <= m_eventTreshold) ||
          (remainingDays < 0 && remainingDays >= (m_pastTreshold* -1)) ||
                remainingDays == 0)?true:false;
}

int KBirthdayApplet::printEvent(QPainter* painter, int x, int y, int width, const BListEntry* entry)
{
    int remainingDays = entry->remainingDays();

    if ( !testThreshold(remainingDays))
        return y-3;

    painter->save();

    if (entry->remainingDays() > m_highlightTreshold) {
        painter->setPen(m_colour);
    } else if (remainingDays < 0) {
        painter->setPen(m_past_colour);
    } else {
        painter->setPen(m_highlight_colour);
    }

    painter->drawText(x, y, width, painter->fontMetrics().height(), Qt::AlignLeft,
                 QString("%1").arg(entry->name()));
    painter->drawText(x, y, width, painter->fontMetrics().height(), Qt::AlignRight,
                 QString("%1").arg(entry->age()));

    y += painter->fontMetrics().height();
    
    painter->drawText(x, y, width, painter->fontMetrics().height(), 
                      Qt::AlignRight, 
                      remainingDaysString(remainingDays));

    painter->restore();

    return y+painter->fontMetrics().height();
}

QString KBirthdayApplet::remainingDaysString(const int remainingDays)
{
    QString msg;
    if (remainingDays < -1) {
        msg = i18n("%1 days ago",remainingDays*-1);
    } else if (remainingDays == -1) {
        msg = i18n("yesterday");
    } else if (remainingDays == 0) {
        msg = i18n("today");
    } else if (remainingDays == 1) {
        msg = i18n("tomorrow");
    } else {
        msg = i18n("in %1 days",remainingDays);
    }
    return msg;
}

#include "kbirthdayapplet.moc"
